frame/contracts/src/exec.rs

```diff
@@ -1152,11 +1168,12 @@ where
 		let contract_info = top_frame.contract_info().clone();
 		let account_id = top_frame.account_id.clone();
 		let value = top_frame.value_transferred;
+		let caller = self.caller().account_id()?;
 		let executable = self.push_frame(
 			FrameArgs::Call {
 				dest: account_id,
 				cached_info: Some(contract_info),
-				delegated_call: Some(DelegatedCall { executable, caller: self.caller().clone() }),
+				delegated_call: Some(DelegatedCall { executable, caller: caller.clone() }),
```

Comment:
We should allow root here. Otherwise a contract called by root can't use delegate call.